File: app/src/main/java/fr/insapp/insapp/adapters/UserRecyclerViewAdapter.java
package fr.insapp.insapp.adapters;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.bumptech.glide.RequestManager;

import java.util.ArrayList;
import java.util.List;

import de.hdodenhof.circleimageview.CircleImageView;
import fr.insapp.insapp.R;
import fr.insapp.insapp.models.User;
import fr.insapp.insapp.utility.Utils;

import static com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions.withCrossFade;

/**
 * Created by thomas on 24/02/2017.
 */

public class UserRecyclerViewAdapter extends BaseRecyclerViewAdapter<UserRecyclerViewAdapter.UserViewHolder> {

    private RequestManager requestManager;

    private boolean matchParent;

    protected List<User> users;

    private OnUserItemClickListener listener;

    public interface OnUserItemClickListener {
        void onUserItemClick(User user);
    }

    public UserRecyclerViewAdapter(Context context, RequestManager requestManager, boolean matchParent) {
        this.context = context;
        this.requestManager = requestManager;
        this.matchParent = matchParent;
        this.users = new ArrayList<>();
    }

    public void setOnItemClickListener(OnUserItemClickListener listener) {
        this.listener = listener;
    }

    public void addItem(User user) {
        this.users.add(user);
        this.notifyDataSetChanged();
    }

    @Override
    public UserRecyclerViewAdapter.UserViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        final View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.user_thumb, parent, false);
        return new UserViewHolder(view, matchParent);
    }

    @Override
    public void onBindViewHolder(UserRecyclerViewAdapter.UserViewHolder holder, int position) {
        final User user = users.get(position);

        // get the drawable of avatarCircleImageView

        final int id = context.getResources().getIdentifier(Utils.INSTANCE.drawableProfileName(user.getPromotion(), user.getGender()), "drawable", context.getPackageName());
        requestManager
                .load(id)
                .transition(withCrossFade())
                .into(holder.avatar);

        holder.name.setText(user.getName());
        holder.username.setText(user.getUsername());

        holder.bind(user, listener);
    }

    @Override
    public int getItemCount() {
        return users.size();
    }

    public List<User> getUsers() { return users; }

    public static class UserViewHolder extends RecyclerView.ViewHolder {
        public CircleImageView avatar;
        public TextView name;
        public TextView username;

        public UserViewHolder(View view, boolean matchParent) {
            super(view);

            if (matchParent)
                (view.findViewById(R.id.user_thumb_macaroon_layout)).getLayoutParams().width = LinearLayout.LayoutParams.MATCH_PARENT;

            this.avatar = (CircleImageView) view.findViewById(R.id.user_avatar);
            this.name = (TextView) view.findViewById(R.id.user_name);
            this.username = (TextView) view.findViewById(R.id.user_username);
        }

        public void bind(final User user, final OnUserItemClickListener listener) {
            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    listener.onUserItemClick(user);
                }
            });
        }
    }
}

File: app/src/main/java/fr/insapp/insapp/adapters/UserRecyclerViewAdapter.kt
package fr.insapp.insapp.adapters

import android.content.Intent
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.RequestManager
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions.withCrossFade
import com.bumptech.glide.request.RequestOptions
import fr.insapp.insapp.R
import fr.insapp.insapp.activities.ProfileActivity
import fr.insapp.insapp.models.User
import fr.insapp.insapp.utility.Utils
import fr.insapp.insapp.utility.inflate
import kotlinx.android.synthetic.main.user_thumb_macaroon.view.*

/**
 * Created by thomas on 24/02/2017.
 * Kotlin rewrite on 28/08/2019.
 */

class UserRecyclerViewAdapter(
        val users: MutableList<User>,
        private val requestManager: RequestManager,
        private val matchParent: Boolean
) : RecyclerView.Adapter<UserRecyclerViewAdapter.UserViewHolder>() {

    fun addItem(user: User) {
        this.users.add(user)
        this.notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) = UserViewHolder(parent.inflate(R.layout.user_thumb), requestManager, matchParent)

    override fun onBindViewHolder(holder: UserViewHolder, position: Int) {
        val user = users[position]
        holder.bindUser(user)
    }

    override fun getItemCount() = users.size

    class UserViewHolder(
            private val view: View,
            private val requestManager: RequestManager,
            matchParent: Boolean
    ) : RecyclerView.ViewHolder(view), View.OnClickListener {

        private var user: User? = null

        init {
            view.setOnClickListener(this)

            if (matchParent)
                view.user_thumb_macaroon_layout.layoutParams.width = LinearLayout.LayoutParams.MATCH_PARENT
        }

        fun bindUser(user: User) {
            this.user = user

            val context = itemView.context

            val id = context.resources.getIdentifier(Utils.drawableProfileName(user.promotion, user.gender), "drawable", context.packageName)
            requestManager
                .load(id)
                .apply(RequestOptions.circleCropTransform())
                .transition(withCrossFade())
                .into(view.user_avatar)

            view.user_name.text = user.name
            view.user_username.text = user.username
        }

        override fun onClick(v: View) {
            val context = itemView.context
            context.startActivity(Intent(context, ProfileActivity::class.java).putExtra("user", user))
        }
    }
}
